src/job_market.rs

```diff
@@ -46,20 +51,44 @@ struct JobMarket<Job> {
     job_batches: Vec<VecDeque<Job>>,
 }
 
-impl<Job> JobBroker<Job> {
+impl<Job> JobBroker<Job>
+where
+    Job: Send + 'static,
+{
     /// Create a new market for a group of threads.
-    pub fn new(thread_count: usize) -> Self {
-        Self {
+    pub fn new(thread_count: usize, close_at: Option<SystemTime>) -> Self {
+        let s = Self {
             has_new_jobs: Arc::new(Condvar::new()),
             market: Arc::new(Mutex::new(JobMarket {
                 open: true,
                 thread_count,
                 open_count: thread_count,
                 job_batches: Vec::new(),
             })),
+        };
+        if let Some(closing_time) = close_at {
+            let s1 = s.clone();
+            std::thread::Builder::new()
+                .name("timeout".to_owned())
+                .spawn(move || loop {
+                    let mut market = s1.market.lock();
+                    let now = SystemTime::now();
+                    if closing_time < now {
+                        log::debug!("Reached timeout, triggering shutdown");
+                        market.open = false;
+                    }
+                    if !market.open {
+                        break;
+                    }
+                    sleep(Duration::from_secs(1));
+                })
```

Comment:
Potential future change: could sleep the delta at the beginning of the loop to avoid a wake-up every second.

Reply:
Definitely an improvement: https://github.com/stateright/stateright/pull/77